nova_vm/src/ecmascript/execution/agent.rs

```diff
@@ -52,6 +52,31 @@ pub trait HostHooks: std::fmt::Debug {
     fn host_has_source_text_available(&self, func: Function) -> bool;
 }
 
+pub struct BoxedAgent {
+    agent: Box<Agent>,
+    root_realms: Vec<RealmIdentifier>,
+}
+
+impl BoxedAgent {
+    pub fn new(options: Options, host_hooks: &'static dyn HostHooks) -> Self {
+        Self {
+            agent: Box::new(Agent::new(options, host_hooks)),
+            root_realms: Vec::with_capacity(1),
+        }
+    }
+
+    pub fn with<'agent, F, R>(&'agent mut self, func: F) -> R
+    where
+        F: FnOnce(&'agent mut Agent, &'agent mut Vec<RealmIdentifier>) -> R,
+    {
+        func(&mut self.agent, &mut self.root_realms)
```

Comment:
**issue:** How does this interact with the execution context stack? The stack need not be empty after the closure – in fact, the `nova_vm` public API currently lets you push an execution context with `initialize_host_defined_realm` but not pop out of it. Since execution contexts can contain indices, they might not be being tracked as roots. On the other hand, all execution contexts in the stack at that point should only have realm indices.

**suggestion:** Rather than letting the closure modify the set of root realms, how about using the execution context stack to determine them automatically?

Reply:
Yeah, I kind of looked at it and decided that the execution context stack should at only contain "empty" contexts that just point at the realm.

But you're actually fully correct that they should still be visited and sweeped, in case eg. Realm 2 is GC'd and Realm 3 takes its "place". That cannot happen at present, but maybe in the future.

Reply:
Using the execution context stack does seem like a fairly good idea. I'm a bit worried if that won't lead to us not being able to GC realms that still have eg. forever-pending Promises on them? But that's a problem for another day.

Reply:
I refactored this to use `RealmRoot` items that internally point to the `root_realms` vector. The user no longer has direct access to the vector, and control of realm rooting is done through APIs. The `with` API is now `run_in_realm` which prepares the first execution context item and pops it off at the end.